### PR TITLE
Log the environment

### PR DESCRIPTION
**Pull request by vext01**
Not ready to merge.

Laurie asked last week if we could log the environment of each process execution, so that we could check that each benchmark with the same key (so that's vm + benchmark pairings) ran with the same environment.

This PR makes a start on that. I'm using a line of shell script, as it's the innermost place before the VM itself is invoked.

Questions:
 * Should this only be done when the instrument flag is on?
 * This change introduces I/O just before the benchmark. Should we move the sync wait into the wrapper script (after the new I/O) in the form of  `sync && sleep N`?

If the general approach is OK, then I will need to do some magic with the log file locations. I couldn't write into the experiment dir, as the krun user does not have access. At the moment, I write to /tmp. Likley a good idea is to write first to /tmp, then have krun move it after the execution finishes.

FWIW, I don't see any environment discrepancies using just the example config file. A typical env looks like this:
```
SUDO_GID=0                                                                      
MAIL=/var/mail/krun                                                             
USER=krun                                                                       
HOME=/home/krun                                                                 
SUDO_UID=0                                                                      
LOGNAME=krun                                                                    
TERM=unknown                                                                    
USERNAME=krun                                                                   
PATH=/usr/bin:/bin:/usr/sbin:/sbin:/usr/X11R6/bin:/usr/local/bin                
SUDO_COMMAND=/usr/local/bin/dash /tmp/krun_wrapper.dash                         
SHELL=/bin/ksh                                                                  
SUDO_USER=root                                                                  
PWD=/home/edd/krun/examples                                                     
```

### CODE REVIEW

**Comment by ltratt** [comment]
OK, this looks like a good start.

We could have the parent process pass a first argument to the shell script which is temporary directory, so that we don't have any race conditions? Then we can log all of them easily.

I have two thoughts. Either we move the sync and wait into this shell script *or* this script simply commits the environment to an internal buffer, runs the benchmarks, and *then* writes the file out. That latter approach might be easier?

**Comment by vext01** [comment]
> We could have the parent process pass a first argument to the shell script which is temporary directory, so that we don't have any race conditions? Then we can log all of them easily.

Yep, could do, although I would probably suggest we use the PID of the krun process, as it's easier to implement. Note alsp that some other areas of krun (e.g. the wrapper script filename itself) are guilty of not using unique filenames. I don't think it is "racy" per se though, since you are not supposed to run more than one krun instance anyway. Any thoughts?

> I have two thoughts. Either we move the sync and wait into this shell script or this script simply commits the environment to an internal buffer, runs the benchmarks, and then writes the file out. That latter approach might be easier?

Yes, the latter is easier.

Something like:

```
ENVLOG=`env`
<blah>
echo ${ENVLOG} > file
```

Right?

**Comment by vext01** [comment]
>We could have the parent process pass a first argument to the shell script which is temporary directory, so that we don't have any race conditions? Then we can log all of them easily.

Ah wait. Depending what you mean with this, we might be ok. Note that the env file is derived from the job key and is suffixed with the execution number.

**Comment by vext01** [comment]
Try this.

Afterwards we have have directory (whose name is derived from the config file name) containing env logs, one per exec.

E.g. after running `example.krun`:

```
$ ls example_envlogs/
dummy__CPython__default-python__0.env             dummy__C__default-c__1.env                        nbody__C__default-c__0.env
dummy__CPython__default-python__1.env             nbody__CPython__default-python__0.env             nbody__C__default-c__1.env
dummy__C__default-c__0.env                        nbody__CPython__default-python__1.env
```

**Comment by vext01** [comment]
(Will fix tests)

**Comment by vext01** [comment]
Not sure why travis is taking so long here. It's been over an hour...

**Comment by vext01** [comment]
https://www.traviscistatus.com/incidents/m41c9md5dg9z

**Comment by ltratt** [comment]
Shall we restart the tests or wait?

**Comment by vext01** [comment]
Travis is fixed, but has a backlog. Let's just wait.

**Comment by ltratt** [comment]
OK.

**Comment by vext01** [comment]
Feel free to review though.

File: krun.py [hunk 1]
```diff
@@ -220,6 +223,10 @@ def inner_main(mailer, on_first_invocation, config, args):
         util.fatal("Instrumentation dir '%s' exists (from an older run?)" %
                    (instr_dir))
 
+    if envlog_dir_exists and on_first_invocation:
+        util.fatal("Env log dir '%s' exists (from an older run?)" %
```
Comment (ltratt) [inline krun.py]: I think we can just say "Env log dir '%s' already exists. Please remove it."

**Comment by ltratt** [comment]
This looks OK except I really don't like the racy /tmp/-with-PID hack. I know we've done that sort of thing elsewhere in Krun, but it's caused us problems, and I can't see that it's a good plan to repeat the same idiom ;) Why can't we just use `mktemp -d` and pass that name around?

**Comment by vext01** [comment]
Using a PID in the filename is an improvement on what we did in the past (I have a branch to fix this).

But I don't understand how this is racy. A PID is unique and we are single threaded.

**Comment by ltratt** [comment]
Is there a technical reason that we can't use `mktemp -d`? If so, we should do that, because the OS guarantees there can't be any problems. That way we absolve ourselves from worrying about single-threadedness, or existing directories (PIDs can be reused) etc.

**Comment by vext01** [comment]
We can use `mktemp` (why -`d?`), but it requires some plumbing.

**Comment by vext01** [comment]
(whereas it was easy to know the filename to remove later when it uses a PID, we would have to store the filename somplace so we could pick it up later).

**Comment by vext01** [comment]
A slight variation upon you suggestion would be to store our temp files in a unique directory in /tmp, rather than the files themselves with unique names directly in /tmp. This would be easier to plumb (I think).

**Comment by vext01** [comment]
Scratch that, the files we would want to group together are owned by different users and one one the users may not exists yet.

I'll come up with something...

**Comment by ltratt** [comment]
It's fine to create a directory with `mktemp -d` and then chmod it to g+w (or a+w).

**Comment by vext01** [comment]
The user is created and deleted a lot, so that itself may be racy. I have another plan. Watch this space.

File: krun.py (hunk 1 above)
Comment (vext01) [inline krun.py]: I'm going to stick with just "Env log dir '%s' already exists". If this fires, then krun's internal consistency is broken and the user is on their own to decide what to do.

**Comment by vext01** [comment]
I think this is ready for review.

I wanted to make a top-level unique directory under /tmp which we could use a scratch area for temp files, but this is awkward, as the permissions need to be right. The krun user is created fresh many times in one invocation (due to sanity checks), and we have to guarantee that the uid will be the same each time.

I've fallen back to just using uniquely named temp files directly under /tmp. This is free of the above problem. Keep it simple.

If this is good, my next PR is to do the same to the wrapper script filename.

**Comment by vext01** [comment]
[I also learned that group write permission is not enough to redirect into a file owned by someone else, or to remove a file owned by someone else.]

File: krun/vm_defs.py [hunk 2]
```diff
@@ -108,18 +116,40 @@ def run_exec(self, entry_point, benchmark, iterations, param, heap_lim_k,
                  stack_lim_k, force_dir=None, sync_disks=True):
         pass
 
-    @staticmethod
-    def make_wrapper_script(args, heap_lim_k, stack_lim_k):
-        """Make lines for the wrapper script.
+    def make_wrapper_script(self, args, heap_lim_k, stack_lim_k):
+        """Write the wrapper script.
         Separate for testing purposes"""
```
Comment (ltratt) [inline krun/vm_defs.py]: I think we need to add something like "returns a filename for the env file"?

**Comment by ltratt** [comment]
One minor comment only, I think. Looks good.

**Comment by vext01** [comment]
OK, that's fixed.

One question from earlier outstanding. This is always on, right? Or only for instrumentation mode? I think the former is OK due to the way we have implemented it, right?

**Comment by ltratt** [comment]
OK, if you squash, then I can merge.

**Comment by vext01** [comment]
One more minor commit. OK?

**Comment by vext01** [comment]
Argh, I screwed that up. Moment.

**Comment by vext01** [comment]
Try that. Basically, I was confusing myself because group perms came before user perms.

**Comment by ltratt** [comment]
Are there new commits? I'm confused!

**Comment by vext01** [comment]
Yes, but GH has screwed up the order of the comments and commits.

The last two commits are new.

**Comment by ltratt** [comment]
I'm with it! OK, let's squash.

**Comment by vext01** [comment]
Splat.

**Comment by ltratt** [comment]
OK, waiting for tests to pass...

**Pull request merged**